parachain/primitives/ethereum/src/ethashproof.rs

```diff
@@ -0,0 +1,363 @@
+    pub fn hashimoto_merkle(
+        &self,
+        header_hash: H256,
+        nonce: H64,
+        header_number: u64,
+        nodes: &[DoubleNodeWithMerkleProof],
+    ) -> Result<(H256, H256), Error> {
+        // Check that we have the expected number of nodes with proofs
+        const MIXHASHES: usize = MIX_BYTES / HASH_BYTES;
+        if nodes.len() != MIXHASHES * ACCESSES / 2 {
+            return Err(Error::UnexpectedNumberOfNodes);
+        }
+    
+        let epoch = header_number / EPOCH_LENGTH;
+        // Reuse single Merkle root across all the proofs
+        let merkle_root = self.dag_merkle_root(epoch).ok_or(Error::EpochOutOfRange)?;
+        let full_size = ethash::get_full_size(epoch as usize);
+
+        // Boxed index since ethash::hashimoto gets Fn, but not FnMut
+        let index = RefCell::new(0);
+        // Flag for whether the proof is valid
+        let success = RefCell::new(true);
+
+        let results = ethash::hashimoto_with_hasher(
+            header_hash,
+            nonce,
+            full_size,
+            |offset| {
+                let idx = *index.borrow_mut();
+                *index.borrow_mut() += 1;
+
+                // Each two nodes are packed into single 128 bytes with Merkle proof
+                let node = &nodes[idx / 2];
+                if idx % 2 == 0 {
+                    // Divide by 2 to adjust offset for 64-byte words instead of 128-byte
+                    if let Ok(computed_root) = node.apply_merkle_proof((offset / 2) as u64) {
+                        if merkle_root != computed_root {
+                            success.replace(false);
+                        }
+                    } else {
+                        success.replace(false);
+                    }
+                };
+
+                // Reverse each 32 bytes for ETHASH compatibility
+                let mut data = node.dag_nodes[idx % 2].0;
+                data[..32].reverse();
+                data[32..].reverse();
+                data.into()
+            },
+            keccak_256,
+            keccak_512,
+        );
+
+        match success.into_inner() {
+            true => Ok(results),
+            false => Err(Error::InvalidMerkleProof),
+        }
+    }
+
+    pub fn hashimoto_light(
+        &mut self,
+        header_hash: H256,
+        nonce: H64,
+        header_number: u64,
+    ) -> (H256, H256) { 
+        let epoch = header_number / EPOCH_LENGTH;
+        let cache = match self.dags_cache {
+            Some(ref mut c) => c.get(epoch, header_number),
+            None => panic!("EthashProver wasn't configured with hashimoto light cache"),
+        };
+        let full_size = ethash::get_full_size(epoch as usize);
+        return ethash::hashimoto_light(header_hash, nonce, full_size, cache.as_slice());
+    }
+}
+
+// https://github.com/paritytech/substrate/commit/510e68b8d06a3d407eda0d4c1c330bd484140b65
+fn keccak_512(data: &[u8]) -> [u8; 64] {
```

Comment:
I think this should rather be a host function (`sp_io::hashing`) for performance reasons.

Reply:
I left a comment in the linked PR, cause I think it was an oversight to not include it there.

Reply:
Cool, will add an issue to our repo here and please keep us updated when/if this host function is added for us to use

Reply:
Added a PR here https://github.com/paritytech/substrate/pull/7531